File: algorithm/python/graph/bfs.py
# -*-encoding:utf-8-*-
from tracer import *
logger = LogTracer()
# var tracer = new DirectedGraphTracer().attach(new LogTracer());
tracer = DirectedGraghTracer.attach(logger.getID())
G = [ # G[i][j] indicates whether the path from the i-th node to the j-th node exists or not
    [0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
]
tracer.setData(G, 0)


def bfs(s):
    Q = []
    Q.append(s)
    l = 0
    tracer.visit(s).wait()
    while l < len(Q):
        u = Q[l]
        for i in range(len(G[u])):
            if G[u][i]:
                Q.append(i)
                tracer.visit(i, u).wait()
        l += 1


if __name__ == '__main__':
    bfs(0)
    tracerlist.info()

File: algorithm/javascript/graph/bfs.js
var tracer = new DirectedGraphTracer().attach(new LogTracer());
var G = [ // G[i][j] indicates whether the path from the i-th node to the j-th node exists or not
    [0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
];
tracer._setTreeData(G, 0);


function BFS(s) { // s = start node
    var Q = [];
    Q.push(s); // add start node to queue
    tracer._visit(s)._wait();
    while (Q.length > 0) {
        var node = Q.shift(); // dequeue
        for (var i = 0; i < G[node].length; i++) {
            if (G[node][i]) { // if current node has the i-th node as a child
                Q.push(i); // add child node to queue
                tracer._visit(i, node)._wait();
            }
        }
    }
}
BFS(0);